libs/sdk-ui-dashboard/src/model/store/filterContext/filterContextReducers.ts

```diff
@@ -352,6 +361,153 @@ const clearAttributeFiltersSelection: FilterContextReducer<
+    readonly filterLocalId: string;
+}
+
+const updateFiltersToIndexMappingOnFilterAdded: FilterContextReducer<PayloadAction<string>> = (
+    state,
+    action,
+) => {
+    invariant(state.filtersToIndexMap, "Attempt to edit uninitialized filter context");
+
+    const filterLocalId = action.payload;
+
+    const lastIndex = max(Object.values(state.filtersToIndexMap)) || -1;
+
+    state.filtersToIndexMap[filterLocalId] = lastIndex + 1;
+};
+
+export interface IUpdateFiltersToIndexMappingOnFilterDeletedPayload {
+    readonly filterLocalId: string;
+}
+
+const updateFiltersToIndexMappingOnFilterDeleted: FilterContextReducer<PayloadAction<string>> = (
+    state,
+    action,
+) => {
+    invariant(state.filtersToIndexMap, "Attempt to edit uninitialized filter context");
+
+    const filterLocalId = action.payload;
+
+    delete state.filtersToIndexMap[filterLocalId];
+};
+
+export interface IUpdateConnectingAttributesOnFilterAddedPayload {
+    addedFilterLocalId: string;
+    connectingAttributes: IParentWithConnectingAttributes[];
+}
+
+const updateConnectingAttributesOnFilterAdded: FilterContextReducer<
+    PayloadAction<IUpdateConnectingAttributesOnFilterAddedPayload>
+> = (state, action) => {
+    const { addedFilterLocalId, connectingAttributes } = action.payload;
+    const addedFilterIndex = state.filtersToIndexMap?.[addedFilterLocalId];
+
+    const connectingAttributeMatrix = cloneDeep(state.connectingAttributeMatrix);
+
+    invariant(
+        addedFilterIndex !== undefined && connectingAttributeMatrix,
+        "Attempt to edit uninitialized or incorrectly initialized filter context",
+    );
+
+    connectingAttributeMatrix[addedFilterIndex] = [];
+
+    for (const connectingAttribute of connectingAttributes) {
+        const neighborFilterIndex = state.filtersToIndexMap?.[connectingAttribute.filterLocalId];
+
+        invariant(
+            neighborFilterIndex !== undefined,
+            "Attempt to edit uninitialized or incorrectly initialized filter context",
+        );
+
+        connectingAttributeMatrix[neighborFilterIndex][addedFilterIndex] =
+            connectingAttribute.connectingAttributes;
+        connectingAttributeMatrix[addedFilterIndex][neighborFilterIndex] =
+            connectingAttribute.connectingAttributes;
+    }
+
+    state.connectingAttributeMatrix = connectingAttributeMatrix;
+};
+
+const updateConnectingAttributesOnFilterDeleted: FilterContextReducer<PayloadAction<string>> = (
+    state,
+    action,
+) => {
+    const deletedFilterLocalId = action.payload;
+    const deletedFilterIndex = state.filtersToIndexMap?.[deletedFilterLocalId];
+    const connectingAttributeMatrix = cloneDeep(state.connectingAttributeMatrix);
+
+    invariant(
+        deletedFilterIndex !== undefined && connectingAttributeMatrix,
+        "Attempt to edit uninitialized or incorrectly initialized filter context",
+    );
+
+    for (let index = 0; index < connectingAttributeMatrix.length; index++) {
```

Comment:
Written like this, the matrix will can grow unboundedly. If you add a filter, the remove another filter and so on, this will grow O(n^2). Can you instead delete the row and the column and fix the indexes in the index map?

Reply:
On delete, the empty array on deleted index remains. But when adding another filter, the new attribute filter is assigned to the lowest possible index.